polus-image-registration-plugin/src/image_registration.py

```diff
@@ -0,0 +1,477 @@
+        # remove all coordinates which are out of bounds (negative coordinates, or that which lie outside the moving image)
+        boo = (np.all(new_array>=0, axis =0)) * (new_array[0] <width_2) * (new_array[1] <height_2)
+        new_array=new_array[:,boo]
+        row_array=row_array[:,boo]
+        # fetch pixel values from the moving image and place them at their transformed position in the transformed image
+        transformed_image[row_array[1,:], row_array[0,:]]= image[new_array[1,:], new_array[0,:]]    
+    return transformed_image
+
+def get_tile_by_tile_transformation(reference_image_tiles, moving_image_tiles, scale_matrix): 
+    """
+    This function takes as input a set of tiles from the reference image and moving image and
+    calculates the homography tranformation between respective tiles. The transformation is
+    calculated between scaled down versions of the images, the scale matrix is used to upscale
+    the homography matrixes.
+    
+    Input:
+        reference_image_tiles= list of tiles from the reference image
+        moving_image_tiles= list of tiles from the moving image
+        scale_matrix= used to up_scale the homography matrices 
+        
+    Output:
+        homography_set = list of homography matrixes 
+        
+    """   
+    NUMBER_OF_TILES=4
+    homography_set=[]       
+    for k in range(NUMBER_OF_TILES):
+        # cacluate transformation betwen corresponding tiles
+        _ , homography_matrix=image_transformation(moving_image_tiles[k], reference_image_tiles[k] )
+        # upscale and append the matrix
+        homography_set.append(homography_matrix*scale_matrix)            
+    return homography_set
+
+
+def apply_registration(moving_image_path,Template_image_shape,Rough_Homography_Upscaled, fine_homography_set):
+    
+    """
+    This function transforms an image using the set of transformations (rough and fine) returned
+    by the register_images() function below. 
+    
+    Inputs: 
+        moving_image_path: path to the moving image
+        Template_image_shape: dimensions of the reference image
+        Rough_Homography_Upscaled: rough homography matrix returned by register_images() function
+        fine_homography_set: set of tile by tile transformation matrices returned by register_images(function)
+        
+    Outputs:
+        transformed_moving_image: transformed moving image
+        moving_image_metadata: metadata from the moving image       
+        
+    """
+    
+    logging.basicConfig(format='%(asctime)s - %(name)-8s - %(levelname)-8s - %(message)s',
+                        datefmt='%d-%b-%y %H:%M:%S')
+    logger = logging.getLogger("apply_registration()")
+    logger.setLevel(logging.INFO)
+    
+    # read image using bfio
+    bf = BioReader(moving_image_path)
+    moving_image_metadata=bf.read_metadata()
+    moving_image = bf.read_image(C=[0])    
+    moving_image=moving_image[:,:,0,0,0] 
+    del bf
+    
+    mem = psutil.virtual_memory()
```

Comment:
You should consider adding an if statement around the usage of psutil to only track memory under debugging circumstances. Also, the logger should probably be `logger.debug` since I don't think this type of information really helps the end user understand the progress of the plugin execution.

This comment applies to the rest of the code that contains debugging code. Please revise everywhere psutil is used. Also, review the usage of the logger and only include info messages that give execution status updates and any debugging code should be switched to debug.

Reply:
Usage of psutil has been limited to only provide memory stats at the beginning of execution.